orttraining/orttraining/training_ops/cuda/controlflow/wait.cc

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/tensor/utils.h"
+// Include RecordEvent's utility functions shared by CPU and GPU implementations.
+#include "orttraining/training_ops/cpu/controlflow/common.h"
+// Include event mechanism shared by CPU and GPU implementations.
+#include "orttraining/training_ops/cpu/controlflow/event_pool.h"
+#include "wait.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+    WaitEvent,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
```

Comment:
>InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */ [](start = 9, length = 59)

Seems this kernel declaration is the only one that's different from CPU implementation.

Reply:
Another difference is CopyTensor v.s. CopyCpuTensor.

Reply:
But if CopyTensor does shortcut when X and Y are the same, then these code can share right?

---
In reply to: [413971952](https://github.com/microsoft/onnxruntime/pull/3653#discussion_r413971952) [](ancestors = 413971952)

Reply:
It's not safe. We can't do function overloading to distinguish the call to GPU and the call to CPU function.

Reply:
I have merged the rest part in CPU and GPU implementations in the last commit.